scripts/gen_release_notes/tests/test_file_writer.py

```diff
@@ -0,0 +1,61 @@
+"""
+Test function to check if the methods declared in file_writer work as intended
+"""
+
+import os
+from unittest.mock import Mock
+
+import pytest
+from gen_release_notes.file_writer import file_writer
+
+sample_changelog_list = [
+    {
+        "dev": [
+            {
+                "type_change": "doc",
+                "change_entries": [
+                    "env/dev/someproject/something/nice.tf",
+                    "docs/something/data.md",
+                    "docs/something/another.md",
+                ],
+            },
+            {
+                "type_change": "awslambda",
+                "change_entries": ["awslambda/anotheroot/changes.py", "awslamda/anothertest/changes.py"],
+            },
+        ]
+    }
+]
+
+sample_dir_list = [
+    ["dev"],
+    ["dev", "tst"],
+]
+
+sample_cli_args = [
+    {"next_release_tag_branch": "some_bracnch", "sprint_number": "21"},
+    {"next_release_tag_branch": "another_branch", "sprint_number": "31"},
+]
+
+output_path = "sample.md"
+
+
+@pytest.mark.parametrize("user_input_changelog_obj", sample_changelog_list)
+@pytest.mark.parametrize("user_input_dir_list", sample_dir_list)
+@pytest.mark.parametrize("user_input_cli_args", sample_cli_args)
+def test_write_to_file(user_input_changelog_obj, user_input_dir_list, user_input_cli_args):
+    """
+    Test to check if the file generation works as intended.
+
+    Asserts if we created a file that has contents in it
+    """
+
+    mock_env = Mock()
+    mock_env.next_release_tag_branch = Mock(return_value=user_input_cli_args["next_release_tag_branch"])
+    mock_env.sprint_number = Mock(return_value=user_input_cli_args["sprint_number"])
+
+    file_writer.write_contents_to_file(user_input_changelog_obj, user_input_dir_list, mock_env, output_path)
```

Comment:
I would prefer we mock the writing of the object to disk so that way we dont have to call `os.remove`. @davido-nw might have some good tips for doing this but if we want to do this later im ok with that too.

Reply:
This might be useful: https://docs.python.org/3/library/unittest.mock.html#mock-open

Reply:
Also - check out the test cases I wrote for the AWS key rotator, since that's a script that involves writing to a file. In particular, this pattern where `mock_open` instantiates a "mock file" with data, then the patch to `builtins.open` makes all `open` syscalls use that mock file:
```py
fake_creds_fp = mock_open(read_data=creds_data)
with patch("builtins.open", fake_creds_fp):
    rotator.IAMKeyRotator(config).main()
````

Reply:
thanks @davido-nw for passing this along

Reply:
Thanks for the help on this. I almost got it to work with this instead; I'm currently running into an issue with the `mock_open` also patching out one of the jinja2 methods, which is causing the following errors to occur:

```
________________________________________ test_write_to_file[user_input_cli_args1-user_input_dir_list1-user_input_changelog_obj0] ________________________________________

user_input_changelog_obj = {'dev': [{'change_entries': ['env/dev/someproject/something/nice.tf', 'docs/something/data.md', 'docs/something/anothe...'change_entries': ['awslambda/anotheroot/changes.py', 'awslamda/anothertest/changes.py'], 'type_change': 'awslambda'}]}
user_input_dir_list = ['dev', 'tst'], user_input_cli_args = {'next_release_tag_branch': 'another_branch', 'sprint_number': '31'}

    @pytest.mark.parametrize("user_input_changelog_obj", sample_changelog_list)
    @pytest.mark.parametrize("user_input_dir_list", sample_dir_list)
    @pytest.mark.parametrize("user_input_cli_args", sample_cli_args)
    def test_write_to_file(user_input_changelog_obj, user_input_dir_list, user_input_cli_args):
        """
        Test to check if the file generation works as intended.

        Asserts if we created a file that has contents in it
        """

        mock_env = Mock()
        mock_env.next_release_tag_branch = Mock(return_value=user_input_cli_args["next_release_tag_branch"])
        mock_env.sprint_number = Mock(return_value=user_input_cli_args["sprint_number"])

        with patch("builtins.open", mock_open()) as mocked_file:
>           file_writer.write_contents_to_file(user_input_changelog_obj, user_input_dir_list, mock_env, output_path)

tests/test_file_writer.py:57:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
file_writer/file_writer.py:23: in write_contents_to_file
    top_template = jinja_env.get_template("beginning.j2")
.tox/py38/lib/python3.8/site-packages/jinja2/environment.py:997: in get_template
    return self._load_template(name, globals)
.tox/py38/lib/python3.8/site-packages/jinja2/environment.py:958: in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
.tox/py38/lib/python3.8/site-packages/jinja2/loaders.py:125: in load
    source, filename, uptodate = self.get_source(environment, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <jinja2.loaders.FileSystemLoader object at 0x102df0190>, environment = <jinja2.environment.Environment object at 0x102df0130>, template = 'beginning.j2'

    def get_source(
        self, environment: "Environment", template: str
    ) -> t.Tuple[str, str, t.Callable[[], bool]]:
        pieces = split_template_path(template)
        for searchpath in self.searchpath:
            filename = os.path.join(searchpath, *pieces)
            f = open_if_exists(filename)
            if f is None:
                continue
            try:
>               contents = f.read().decode(self.encoding)
E               AttributeError: 'str' object has no attribute 'decode'

.tox/py38/lib/python3.8/site-packages/jinja2/loaders.py:201: AttributeError
```

I think I would just need to patch out the jinja2 functions I'm using in this case so that the patch is only patching the method I'm interested in.